src/core/server/saved_objects/migrationsv2/index.ts

```diff
@@ -0,0 +1,857 @@
+          // installed / enabled.
+          controlState: 'UPDATE_TARGET_MAPPINGS',
+          // Source is a none because we didn't do any migration from a source
+          // index
+          source: Option.none,
+          target: `${stateP.indexPrefix}_${stateP.kibanaVersion}_001`,
+          targetMappings: mergeMigrationMappingPropertyHashes(
+            stateP.targetMappings,
+            indices[aliases[currentAlias(stateP)]].mappings
+          ),
+          versionIndexReadyActions: Option.none,
+        };
+      } else if (
+        // `.kibana` is pointing to an index that belongs to a later
+        // version of Kibana .e.g. a 7.11.0 node found `.kibana_7.12.0_001`
+        valid(indexVersion(aliases[currentAlias(stateP)])) &&
+        gt(indexVersion(aliases[currentAlias(stateP)]), stateP.kibanaVersion)
+      ) {
+        stateP = {
+          ...stateP,
+          controlState: 'FATAL',
+          error: new Error(
+            `The ${currentAlias(
+              stateP
+            )} alias is pointing to a newer version of Kibana: v${indexVersion(
+              aliases[currentAlias(stateP)]
+            )}`
+          ),
+        };
+      } else if (
+        // If the `.kibana` alias exists
+        aliases[currentAlias(stateP)] != null
+      ) {
+        // The source index is the index the `.kibana` alias points to
+        const source = aliases[currentAlias(stateP)];
+        const target = versionIndex(stateP);
+        stateP = {
+          ...stateP,
+          controlState: 'SET_SOURCE_WRITE_BLOCK',
+          source: Option.some(source) as Option.Some<string>,
+          target,
+          targetMappings: mergeMigrationMappingPropertyHashes(
+            stateP.targetMappings,
+            indices[source].mappings
+          ),
+          versionIndexReadyActions: Option.some<AliasAction[]>([
+            { remove: { index: source, alias: currentAlias(stateP), must_exist: true } },
+            { add: { index: target, alias: currentAlias(stateP) } },
+            { add: { index: target, alias: versionAlias(stateP) } },
+          ]),
+        };
+      } else if (indices[legacyIndex(stateP)] != null) {
+        // Migrate from a legacy index
+
+        // If the user used default index names we can narrow the version
+        // number we use when creating a backup index. This is purely to help
+        // users more easily identify how "old" and index is so that they can
+        // decide if it's safe to delete these rollback backups. Because
+        // backups are kept for rollback, a version number is more useful than
+        // a date.
+        let legacyVersion = '';
+        if (stateP.indexPrefix === '.kibana') {
+          legacyVersion = 'pre6.5.0';
+        } else if (stateP.indexPrefix === '.kibana_task_manager') {
+          legacyVersion = 'pre7.4.0';
+        } else {
+          legacyVersion = 'pre' + stateP.kibanaVersion;
+        }
```

Comment:
`kibana.index: ".my-kibana"` or a new saved object index will be reindexed into `.my-kibana_pre7.11.0_001` because we couldn't identify what type of index this was (and therefore what version of kibana it's from). But I need to add tests for these cases.

I've been wondering if we should let plugins define which indices they want to use. Having a dedicated index for your plugin is rather unusual so it feels like something Core should add to an allowlist (and then a plugin can choose between any indices in the allowlist). If we do this we could add the extra meta information of what version of Kibana introduced this index. It would require more changes than what I'd like to do in this PR though.

I'm wondering if we shouldn't just always use something like `.my-kibana_pre7.11.0_backup_001` then the pattern is consistent no matter what your `kibana.index` or `task_manager.index` config settings are and it's still useful for helping you decide if it's safe to delete this index. Cause ultimately we're just trying to be helpful in choosing a name that's less cryptic than `.kibana_10` so that you know "my upgrade to 7.11 succeeded and I've verified everything so I'm confident I can remove this data now".